tests/Database/DatabaseEloquentSoftDeletesIntegrationTest.php

```diff
@@ -878,14 +879,32 @@ public function testMorphToNonSoftDeletingModel()
         $this->assertNull($comment->owner);
     }
 
+    public function testSelfReferencingRelationshipWithSoftDeletes()
+    {
+        /*
+         * https://github.com/laravel/framework/issues/42075
+         */
+        [$taylor, $abigail] = $this->createUsers();
+
+        $versionsA = $abigail->versions;
+        $versionsT = $taylor->versions;
+
+        self::assertCount(1, $versionsA);
+        self::assertTrue($versionsA->first()->is($taylor));
+
+        self::assertCount(0, $versionsT);
+
+        self::assertEquals(1, SoftDeletesTestUser::whereHas('versions')->count());
+    }
```

Comment:
Would match the code style of the rest of the tests (by using `$this` over `self::` and inline your variables `$versionsA` and `$versionsT`. More like this:

```php
    public function testSelfReferencingRelationshipWithSoftDeletes()
    {
        [$taylor, $abigail] = $this->createUsers();

        $this->assertCount(1, $abigail->versions);
        $this->assertTrue($abigail->versions->first()->is($taylor));

        $this->assertCount(0, $taylor->versions);
        $this->assertEquals(1, SoftDeletesTestUser::whereHas('versions')->count());
    }
```

There also may be a better name than `version` for this "test" relationship. Maybe `self_referencing_relationships()` to make the test relation match the test case, making it more clear?

    public function testSelfReferencingRelationshipWithSoftDeletes()
    {
        [$taylor, $abigail] = $this->createUsers();

        $this->assertCount(1, $abigail->self_referencing_relationships);
        $this->assertTrue($abigail->self_referencing_relationships->first()->is($taylor));

        $this->assertCount(0, $taylor->self_referencing_relationships);
        $this->assertEquals(1, SoftDeletesTestUser::whereHas('self_referencing_relationships')->count());
    }